pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -0,0 +1,318 @@
+	// test; verifying that the changefeed on the `warehouse` table
+	// works is sufficient for our purposes
+	target = "tpcc.warehouse"
+)
+
+func registerCDCMixedVersions(r registry.Registry) {
+	r.Add(registry.TestSpec{
+		Name:            "cdc/mixed-versions",
+		Owner:           registry.OwnerTestEng,
+		Cluster:         r.MakeClusterSpec(5),
+		Timeout:         30 * time.Minute,
+		RequiresLicense: true,
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			runCDCMixedVersions(ctx, t, c, *t.BuildVersion())
+		},
+	})
+}
+
+// cdcMixedVersionTester implements mixed-version/upgrade testing for
+// CDC. It knows how to set up the cluster to run Kafka, monitor
+// events, and run validations that ensure changefeeds work during and
+// after upgrade.
+type cdcMixedVersionTester struct {
+	ctx           context.Context
+	crdbNodes     option.NodeListOption
+	workloadNodes option.NodeListOption
+	kafkaNodes    option.NodeListOption
+	monitor       cluster.Monitor
+	verifierDone  chan struct{}
+	kafka         kafkaManager
+	validator     *cdctest.CountValidator
+	cleanup       func()
+}
+
+func newCDCMixedVersionTester(
+	ctx context.Context, t test.Test, c cluster.Cluster,
+) cdcMixedVersionTester {
+	crdbNodes := c.Range(1, c.Spec().NodeCount-1)
+	lastNode := c.Node(c.Spec().NodeCount)
+
+	c.Put(ctx, t.Cockroach(), "./cockroach", lastNode)
+	c.Put(ctx, t.DeprecatedWorkload(), "./workload", lastNode)
+
+	return cdcMixedVersionTester{
+		ctx:           ctx,
+		crdbNodes:     crdbNodes,
+		workloadNodes: lastNode,
+		kafkaNodes:    lastNode,
+		monitor:       c.NewMonitor(ctx, crdbNodes),
+		verifierDone:  make(chan struct{}),
+	}
+}
+
+// StartKafka will install and start Kafka on the configured node. It
+// will also create the topic where changefeed events will be sent.
+func (cmvt *cdcMixedVersionTester) StartKafka(t test.Test, c cluster.Cluster) {
+	t.Status("starting Kafka node")
+	cmvt.kafka, cmvt.cleanup = setupKafka(cmvt.ctx, t, c, cmvt.kafkaNodes)
+	if err := cmvt.kafka.createTopic(cmvt.ctx, target); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// Cleanup is supposed to be called at the end of tests that use
+// `cdcMixedVersionTester`
+func (cmvt *cdcMixedVersionTester) Cleanup() {
+	if cmvt.cleanup != nil {
+		cmvt.cleanup()
+	}
+}
+
+// installAndStartTPCCWorkload starts a TPCC workload for the given
+// duration. This step does not block; if an error is found while
+// runing the workflow, the test fails.
+func (cmvt *cdcMixedVersionTester) installAndStartTPCCWorkload(d string) versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		t.Status("installing and running workload")
+		tpcc := tpccWorkload{
+			sqlNodes:           cmvt.crdbNodes,
+			workloadNodes:      cmvt.workloadNodes,
+			tpccWarehouseCount: 10,
+			tolerateErrors:     true, // we're bringing nodes down while upgrading
+		}
+
+		tpcc.install(ctx, u.c)
+		time.Sleep(2 * time.Second)
```

Comment:
This seems flaky; what if install takes 5 seconds? will the `run` step time out?

Reply:
That's not the purpose of the `Sleep` call; `install` is synchronous already. This was taken from:

https://github.com/cockroachdb/cockroach/blob/47d34c9dd0c63dfd462e4380a65470f4cf18e3d6/pkg/cmd/roachtest/tests/cdc.go#L171-L173

I should package this into a function to be reused (maybe `install` itself). But before I need to validate that it indeed fails without the `Sleep` call. I'll do that once I have a version of this roachtest running reliably.

Reply:
Yep, that TODO and  error message look quite dated (circa 2018). The `install` imports the fixtures (i.e., pre-computed tpcc tables). I can't think of a reason as to why any `sleep` would be needed since by the time `install` returns, all the tables and their rows would have been propagated down to storage via `AddSSTable`. It would be great to verify and remove it.

Maybe @ajwerner could add some historical context?

Reply:
I think sleep should be dropped.